File: blurb/app/src/main/java/com/mowdowndevelopments/blurb/database/BlurbDao.java
package com.mowdowndevelopments.blurb.database;

import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.Query;

import com.mowdowndevelopments.blurb.database.entities.Feed;
import com.mowdowndevelopments.blurb.database.entities.Story;

import java.util.Collection;
import java.util.List;

@Dao
public interface BlurbDao {

    @Insert
    void addFeeds(Collection<Feed> feeds);

    @Insert
    void addFeed(Feed feed);

    @Insert
    void addStories(Collection<Story> stories);

    @Insert
    void addStory(Story story);

    @Delete
    void removeStory(Story story);

    @Query("SELECT * FROM STORIES")
    LiveData<List<Story>> getFavoriteStories();

    @Query("select * from feeds")
    List<Feed> getFeeds();

    @Query("select feed_title from feeds where id = :id")
    String getFeedTitle(int id);

    @Query("select favicon_url from feeds where id = :id")
    String getFeedFaviconUrl(int id);
}

File: blurb/app/src/main/java/com/mowdowndevelopments/blurb/database/BlurbDao.kt
package com.mowdowndevelopments.blurb.database

import androidx.lifecycle.LiveData
import androidx.paging.DataSource
import androidx.room.*
import com.mowdowndevelopments.blurb.database.entities.Feed
import com.mowdowndevelopments.blurb.database.entities.Story

@Dao
interface BlurbDao {
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun addFeeds(feeds: Collection<Feed>)

    @Insert
    fun addFeed(feed: Feed)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun addStories(stories: Collection<Story>)

    @Insert
    suspend fun addStory(story: Story)

    @Delete
    suspend fun removeStory(story: Story)

    @Query("SELECT * FROM STORIES ORDER BY story_timestamp DESC")
    fun getStarredStoryPagingSourceFactory(): DataSource.Factory<Int, Story>

    @Query("SELECT * FROM STORIES ORDER BY story_timestamp DESC")
    fun getStarredStoryList(): LiveData<List<Story>>

    @Query("SELECT * FROM STORIES ORDER BY story_timestamp DESC LIMIT 20")
    fun getStarredStoryListForWidget(): List<Story>

    @Query("select * from feeds")
    fun getFeeds(): List<Feed>

    @Query("select feed_title from feeds where id = :id")
    fun getFeedTitle(id: Int): String

    @Query("select favicon_url from feeds where id = :id")
    fun getFeedFaviconUrl(id: Int): String

    @Query("select exists(select 1 from stories where story_hash = :storyHash)")
    suspend fun doesStoryExist(storyHash: String?): LiveData<Boolean>
}